### PR TITLE
User Management API with Phone Auth Support

### PR DESCRIPTION
**Pull request by hiranya911**
Adds the basic user management operations to the `auth` module:
* `get_user()`
* `get_user_by_email()`
* `get_user_by_phone_number()`
* `create_user()`
* `update_user()`
* `delete_user()`

Also introduces the following new types for handling user account information:
* `auth.UserRecord`
* `auth.UserInfo`
* `auth.UserMetadata`

go/firebase-python-user-mgt

### CODE REVIEW

File: firebase_admin/auth.py [hunk 1]
```diff
@@ -94,10 +96,653 @@ def verify_id_token(id_token, app=None):
...
+
+    def __init__(self, app):
+        g_credential = app.credential.get_credential()
+        session = transport.requests.AuthorizedSession(g_credential)
+        version_header = 'Python/Admin/{0}'.format(firebase_admin.__version__)
+        session.headers.update({'X-Client-Version': version_header})
+        self._session = session
+
+    def get_user(self, uid):
+        """Gets the user data corresponding to the specified user ID."""
```
Comment (bjornick) [inline firebase_admin/auth.py]: all these get users are basically the same except for the error message.  Does it make sense to have a shared implementation that takes in a parameter for the error messages?

File: firebase_admin/auth.py [hunk 2]
```diff
@@ -94,10 +96,653 @@ def verify_id_token(id_token, app=None):
...
+            if not response or not response.get('users'):
+                raise AuthError(
+                    _UserManager._USER_NOT_FOUND_ERROR,
+                    'No user record found for the provided phone number: {0}'.format(phone_number))
+            return UserRecord(response['users'][0])
+
+    def create_user(self, **kwargs):
+        """Creates a new user account with the specified properties."""
+        payload = {}
+        for key, value in _UserManager._CREATE_USER_FIELDS.items():
```
Comment (bjornick) [inline firebase_admin/auth.py]: I wonder if the kwargs validation/copying logic can be shared between create_user and update_user.

File: firebase_admin/auth.py [hunk 3]
```diff
@@ -14,48 +14,50 @@
```
Comment (bjornick) [inline firebase_admin/auth.py]: Can we split up the contents of this file?  Maybe the validator and the user manager can go in a separate file?

File: tests/test_auth.py [hunk 4]
```diff
@@ -256,3 +261,302 @@ def test_certificate_request_failure(self, authtest):
...
+    assert provider.email is None
+    assert provider.phone_number == '+1234567890'
+    assert provider.display_name is None
+    assert provider.photo_url is None
+    assert provider.provider_id == 'phone'
+
+
+class TestUserRecord(object):
+
+    @pytest.mark.parametrize('data', INVALID_DICTS + [{}, {'foo':'bar'}])
```
Comment (hiranya911) [inline tests/test_auth.py]: Done

File: tests/test_auth.py [hunk 5]
```diff
@@ -256,3 +261,302 @@ def test_certificate_request_failure(self, authtest):
...
+            auth.get_user_by_phone_number(arg)
+
+    def test_get_user_by_phone(self, user_mgt_app):
+        _instrument_user_manager(user_mgt_app, 200, testutils.resource('get_user.json'))
+        _check_user_record(auth.get_user_by_phone_number('+1234567890', user_mgt_app))
+
+    def test_get_user_non_existing(self, user_mgt_app):
+        _instrument_user_manager(user_mgt_app, 200, '{"users":[]}')
+        with pytest.raises(auth.AuthError) as excinfo:
+            auth.get_user('testuser', user_mgt_app)
```
Comment (hiranya911) [inline tests/test_auth.py]: Done

File: tests/test_auth.py [hunk 6]
```diff
@@ -256,3 +261,302 @@ def test_certificate_request_failure(self, authtest):
...
+        _instrument_user_manager(user_mgt_app, 500, '{"error":"test"}')
+        with pytest.raises(auth.AuthError) as excinfo:
+            auth.get_user('testuser', user_mgt_app)
+        assert excinfo.value.code == auth._UserManager._INTERNAL_ERROR
+        assert '{"error":"test"}' in str(excinfo.value)
+
+    def test_get_user_by_email_http_error(self, user_mgt_app):
+        _instrument_user_manager(user_mgt_app, 500, '{"error":"test"}')
+        with pytest.raises(auth.AuthError) as excinfo:
+            auth.get_user_by_email('testuser@example.com', user_mgt_app)
```
Comment (hiranya911) [inline tests/test_auth.py]: Done

File: firebase_admin/auth.py [hunk 7]
```diff
@@ -94,10 +96,653 @@ def verify_id_token(id_token, app=None):
...
+    token_generator = _get_auth_service(app).token_generator
     return token_generator.verify_id_token(id_token)
 
 
+def get_user(uid, app=None):
+    """Gets the user data corresponding to the specified user ID.
+
+    Args:
+        uid: A user ID string.
+        app: An App instance (optional).
```
Comment (hiranya911) [inline firebase_admin/auth.py]: If an app is not explicitly provided (which is the more common use case), we use the default app returned by `firebase_admin.get_app()`. This is a feature common to all APIs in this library. Almost all public functions accept an optional `app` argument. 

The equivalent logic in Java/Android SDKs is:
```
FirebaseAuth.getInstance() // Uses default app
FirebaseAuth.getInstance(app) // Uses the provided app
```

File: firebase_admin/auth.py (hunk 2 above)
Comment (hiranya911) [inline firebase_admin/auth.py]: Done

File: firebase_admin/auth.py [hunk 8]
```diff
@@ -94,10 +96,653 @@ def verify_id_token(id_token, app=None):
...
+
+
+class _Validator(object):
+    """A collectoin of data validation utilities.
+
+    Method provided in this class raise ValueErrors if any validations fail. Normal returns
+    signal success.
+    """
+
+    EMAIL_PATTERN = re.compile('^[^@]+@[^@]+$')
```
Comment (hiranya911) [inline firebase_admin/auth.py]: Changed the validation as suggested. We only need minimal checking here. Backend does more comprehensive checking.

File: firebase_admin/auth.py (hunk 1 above)
Comment (hiranya911) [inline firebase_admin/auth.py]: Done

File: firebase_admin/auth.py (hunk 3 above)
Comment (hiranya911) [inline firebase_admin/auth.py]: Done. Extracted to a new `_user_mgt.py` module. I had to change things around a little bit to break a few cyclic dependencies.

File: tests/test_auth.py [hunk 9]
```diff
@@ -256,3 +261,302 @@ def test_certificate_request_failure(self, authtest):
...
+class TestGetUser(object):
+
+    @pytest.mark.parametrize('arg', INVALID_STRINGS + ['a'*129])
+    def test_invalid_get_user(self, arg):
+        with pytest.raises(ValueError):
+            auth.get_user(arg)
+
+    def test_get_user(self, user_mgt_app):
+        _instrument_user_manager(user_mgt_app, 200, testutils.resource('get_user.json'))
+        _check_user_record(auth.get_user('testuser', user_mgt_app))
```
Comment (hiranya911) [inline tests/test_auth.py]: Done

**Comment by hiranya911** [comment]
Made the suggested changes.

**Comment by dozer47528** [comment]
Hi Firebase team, I would like to know when Firebase team will release this new version include user management APIs ?

**Comment by grennis** [comment]
Will there be a way to lookup multiple users at once? If I have a bunch of user IDs do I need to get their emails one at a time?

**Comment by hiranya911** [comment]
@grennis Not in the initial release of this API. Can you create an issue for multi-user look up? We can consider it for a future iteration (or you can help us implement it ;) )

**Comment by grennis** [comment]
Sure thing. Thanks and looking forward to the release of this PR.

**Pull request merged**